jsoc/gsoc/generativeai.md

```diff
@@ -0,0 +1,107 @@
+- **Suggested Skills and Background**: 
+  - Proficiency in Julia and C programming
+- **Expected Outcomes:** 
+  - Creation of a low-level C interface for llama.cpp
+
+### Project 4: Llama.jl - GPU Support Enhancements
+**Description:** Llama.jl is a Julia interface for [llama.cpp](https://github.com/ggerganov/llama.cpp) that powers many open-source tools today. It's currently leveraging only the high-level binaries. This project focuses on improving the GPU support on non-MacOS platforms (eg, CUDA, ROCm), enabling GPU support for more users.
+
+- **Mentor:** [Mentor's Name and Contact]
+- **Difficulty**: Hard
+- **Duration**: 175 hours
+- **Suggested Skills and Background**: 
+  - Experience with CUDA and GPU computing
+  - Experience with cross-platform development (eg, BinaryBuilder.jl)
+- **Expected Outcomes:** 
+  - Extension of GPU support to additional platforms
+
+### Project 5: AIHelpMe.jl - Enhancing LLM's Julia Knowledge
+
+**Description:** Julia is a low-resource language (ie, not well represented in GenAI training data), so any methods that can bring more precise context or more up-to-date sources are extremely important. AIHelpMe.jl aims to enhance any model's knowledge by injecting the relevant knowledge (eg, Julia code snippets) into the query. The project involves growing the embedded knowledge base and tuning the performance of the Q&A pipeline underpinning all answers.
+
+- **Mentor:** [Jan Siml](https://github.com/svilupp) -- looking for co-mentors?
+- **Difficulty**: Medium
+- **Duration**: 175 hours
+- **Suggested Skills and Background**: 
+  - Familiarity with Julia
+  - Basic understanding of RAG optimization techniques
+- **Expected Outcomes:** 
+  - Expansion of the embedded knowledge base
+  - At least a +10% improvement on the golden Q&A dataset
+
+### Project 6: Native Julia Implementation of ColBERT v2 for Document Retrieval
+
+**Description:** This project involves translating [ColBERT v2](https://github.com/stanford-futuredata/ColBERT), a document retrieval and re-ranking framework, to run natively in Julia. The goal would be to integrate it with AIHelpMe.jl, improving performance and reducing costs.
+
+- **Mentor:** [Jan Siml](https://github.com/svilupp) -- looking for co-mentors?
+- **Difficulty**: Hard
+- **Duration**: 350 hours
+- **Suggested Skills and Background**: 
+  - Familiarity with transformer architecture and Flux.jl (/Transformers.jl)
+  - Experience in document retrieval and re-ranking algorithms
+- **Expected Outcomes:** 
+  - Successful translation of ColBERT v2 to native Julia
+  - Integration with AIHelpMe.jl
+  - Registration of the working package
+
+### Project 7: PromptCompiler.jl - Declarative Interface for GenAI Applications
+
+**Description:** The goal of PromptCompiler.jl is to port [DSPy](https://github.com/stanfordnlp/dspy) to Julia, offering a more user-friendly and efficient way to write complex Generative AI pipelines and programs, leveraging Julia's metaprogramming capabilities.
```

Comment:
```suggestion
**Description:** The goal of this project is to create a new package, PromptCompiler.jl. PromptCompiler.jl would be a port of [DSPy](https://github.com/stanfordnlp/dspy) to Julia, offering a more user-friendly and efficient way to write complex Generative AI pipelines and programs, leveraging Julia's metaprogramming capabilities.
```

Changed this to reflect that this is a greenfield project. I might also consider broadening this out to a more general DSL for prompts, and I think with that we could consider a more general name than PromptCompiler.jl.

Reply:
Fully agreed. But after talking to Jacob, I think it might not be suitable / accepted. It's too broadly scoped atm

Reply:
Agreed. Should we remove the project?

Reply:
Or did you mean the general-purpose DSL?

Reply:
I think there two potentially separate things:
- interface for writing (and orchestrating) GenAI programs (think JuMP) - currently, it would be LangChain, LangGraph, SGLang, etc.
- interface to compile the "prompts" from some sparse directives for a specific task (think Flux with optimization step, ie, it finds its "solution") - this would be represented by DSPy.

I'm much more interested in DSPy-style programming and I intend to add it PromptingTools to experiment with it (or potentially do a small new package from the start)

Reply:
Via Slack, @svilupp and I think this should be pulled.